### ISSUE TITLE
Dataloader that depends on input data

### ISSUE THREAD

**Issue by shoe-diamente** [ChilliCream/hotchocolate#3023]
**Is your feature request related to a problem? Please describe.**
Let's say that we have a type:

```
type Person {
    ...
    posts(isArchived: Bool, take: Int, skip: Int): [Post]
}
```

and then a the query type:

```
type Query {
    people: [Person]
}
```

I would like to be able to use dataloaders in the following query:

```
query {
    people {
        lastArchived: posts(isArchived: true, take: 10) { ... }
        lastNonArchived: posts(isArchived: false, take: 10) { ... }
    }
}
```

**Describe the solution you'd like**
I would like to use dataloaders that depend on resolver's parameters. If possible I'd like to stay away from `IResolverContext`.

I'm not sure if this use case is supported. I don't think so because currently all data loaders don't take into account the fact that a resolver may have additional inputs (like `isArchived`, `take` and `skip`) to the data loader.

**Describe alternatives you've considered**
I'm currently resolving this with these extension methods:

```
    public static class IResolverContextExtensions
    {
        public static IDataLoader<TKey, TValue> BatchDataLoader<TKey, TValue>(
            this IResolverContext context, FetchBatch<TKey, TValue> fetch, string? extraKey = null)
        {
            var loaderKey = Regex.Replace(context.Path.ToString(), @"\[[0-9]+\]", "");
            return context.BatchDataLoader<TKey, TValue>($"{loaderKey}{extraKey}", fetch);
        }

        public static IDataLoader<TKey, TValue> BatchDataLoader<TKey, TValue>(
            this IResolverContext context, FetchBatchCt<TKey, TValue> fetch, string? extraKey = null)
        {
            var loaderKey = Regex.Replace(context.Path.ToString(), @"\[[0-9]+\]", "");
            return context.BatchDataLoader<TKey, TValue>($"{loaderKey}{extraKey}", fetch);
        }

        public static IDataLoader<TKey, TValue[]> GroupDataLoader<TKey, TValue>(
            this IResolverContext context, FetchGroup<TKey, TValue> fetch, string? extraKey = null)
        {
            var loaderKey = Regex.Replace(context.Path.ToString(), @"\[[0-9]+\]", "");
            return context.GroupDataLoader<TKey, TValue>($"{loaderKey}{extraKey}", fetch);
        }

        public static IDataLoader<TKey, TValue[]> GroupDataLoader<TKey, TValue>(
            this IResolverContext context, FetchGroupCt<TKey, TValue> fetch, string? extraKey = null)
        {
            var loaderKey = Regex.Replace(context.Path.ToString(), @"\[[0-9]+\]", "");
            return context.GroupDataLoader<TKey, TValue>($"{loaderKey}{extraKey}", fetch);
        }

        public static IDataLoader<TKey, TValue> CacheDataLoader<TKey, TValue>(
            this IResolverContext context, FetchCache<TKey, TValue> fetch, string? extraKey = null)
        {
            var loaderKey = Regex.Replace(context.Path.ToString(), @"\[[0-9]+\]", "");
            return context.CacheDataLoader<TKey, TValue>($"{loaderKey}{extraKey}", fetch);
        }

        public static IDataLoader<TKey, TValue> CacheDataLoader<TKey, TValue>(
            this IResolverContext context, FetchCacheCt<TKey, TValue> fetch, string? extraKey = null)
        {
            var loaderKey = Regex.Replace(context.Path.ToString(), @"\[[0-9]+\]", "");
            return context.CacheDataLoader<TKey, TValue>($"{loaderKey}{extraKey}", fetch);
        }
    }
```

which basically isolate each resolver to it's own version of the data loader by prefixing the path to the resolver to the data loader key.

But `IResolverContext` is hard to mock correctly in tests so I was trying to move away from `IResolverContext`.

**Additional context**
I'm trying to solve this in HC10 because HC11 is not stable enough for me to use it in production and it's lacking documentation as well.

**Comment by michaelstaib** [comment]
What issue do you have with 11. I would say it is more stable than 10. The docs part I can understand.

**Comment by PascalSenn** [comment]
@shoe-diamente  
I am not sure if I understand this issue :)

You could use a tuple as an data loader argument and then have a member of the tuple for each possible argument, but I am not sure if this helps in you regard

Also why are you using the path as a key for data loaders? This isolates the data loader instance on one path. What do you win with this approach?

**Comment by shoe-diamente** [comment]
> You could use a tuple as an data loader argument and then have a member of the tuple for each possible argument, but I am not sure if this helps in you regard

I tried that but I don't remember what went wrong. I think the dataloader key requires manual serialization into a string or IIRC requires the implementation of `Equals` for each different input type. I'm pretty sure you can't just pass anything to the dataloader key.

> Also why are you using the path as a key for data loaders? This isolates the data loader instance on one path. What do you win with this approach?

As stated what I win is exactly that: isolation of a data loader instance per path (which also implies that the same arguments are passed to that resolver). It's the only way I was able to make it work.

**Comment by shoe-diamente** [comment]
> What issue do you have with 11. I would say it is more stable than 10. The docs part I can understand.

I tried to upgrade a couple of weeks ago but I found it difficult to setup. Especially dataloaders. I don't quite remember the exact annoyances but I remember thinking I didn't agree with the design decisions taken in the newer version.

Then I had another problem with list of objects breaking in errors: https://github.com/ChilliCream/hotchocolate/issues/2947.

But if you tell me HC11 solves the issue above I'll try again.
